### PR TITLE
fix(gateway): fix pathParameter for v2+ api

### PR DESCRIPTION
**Pull request by gcusnieux**
## Issue

https://gravitee.atlassian.net/browse/APIM-446

## Description

A small description of what you did in that PR.

## Additional context

🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.z6.web.core.windows.net/apim-446-fix-path-parameter/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-manbchuihj.chromatic.com)

### CODE REVIEW

File: gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/handlers/api/processor/pathmapping/PathMappingProcessor.java [hunk 1]
```diff
@@ -48,12 +49,25 @@ public void handle(ExecutionContext result) {
...
             .map(Map.Entry::getKey)
-            .min(comparing(o -> countOccurrencesOf(o, ":")))
-            .ifPresent(resolvedMappedPath -> result.request().metrics().setMappedPath(resolvedMappedPath));
+            .min(comparing(o -> countOccurrencesOf(o, PATH_PARAM_PREFIX)))
+            .ifPresent(resolvedMappedPath -> fillMappedPathAndPathParemeters(result, resolvedMappedPath, finalPath));
 
         next.handle(null);
     }
 
+    private void fillMappedPathAndPathParemeters(ExecutionContext result, String resolvedMappedPath, String finalPath) {
```
Comment (ytvnr) [inline gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/handlers/api/processor/pathmapping/PathMappingProcessor.java]:
```suggestion
    private void fillMappedPathAndPathParameters(ExecutionContext result, String resolvedMappedPath, String finalPath) {
```

**Review by ytvnr — APPROVED** [review]

File: gravitee-apim-gateway/gravitee-apim-gateway-flow/src/main/java/io/gravitee/gateway/flow/AbstractFlowParametersProvider.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.*;
+
+public abstract class AbstractFlowParametersProvider implements FlowProvider {
+
+    private static final String URL_PATH_SEPARATOR = "/";
+    private static final String PATH_PARAM_PREFIX = ":";
+
+    public void addContextRequestPathParameters(ExecutionContext context, Flow flow) {
+        Map<Integer, String> parameters = new HashMap();
+        String[] branches = flow.getPath().split(URL_PATH_SEPARATOR);
```
Comment (gcusnieux) [inline gravitee-apim-gateway/gravitee-apim-gateway-flow/src/main/java/io/gravitee/gateway/flow/AbstractFlowParametersProvider.java]: Use pattern

File: gravitee-apim-definition/gravitee-apim-definition-model/src/main/java/io/gravitee/definition/model/v4/listener/http/HttpListener.java [hunk 3]
```diff
@@ -63,7 +63,13 @@ public HttpListener() {
     public void setPathMappings(final Set<String> pathMappings) {
         this.pathMappings = pathMappings;
         if (pathMappings != null) {
-            setPathMappingsPattern(pathMappings.stream().collect(toMap(pathMapping -> pathMapping, Pattern::compile)));
+            setPathMappingsPattern(
+                pathMappings
+                    .stream()
+                    .collect(
+                        toMap(pathMapping -> pathMapping, pathMapping -> Pattern.compile(pathMapping.replaceAll(":\\w*", "[^\\/]*") + "/*"))
```
Comment (jhaeyaert) [inline gravitee-apim-definition/gravitee-apim-definition-model/src/main/java/io/gravitee/definition/model/v4/listener/http/HttpListener.java:70]: I'm not sure to understand the `pathMapping.replaceAll(":\\w*", "[^\\/]*") + "/*")`

Is it made to replace such path `/:id/example/:foo/:bar/something` with a corresponding pattern that matches it? 
If so, I don't understand well the last part with the ` + "/*"` which expects 0 or n '/' ? 🤔

File: gravitee-apim-gateway/gravitee-apim-gateway-flow/src/main/java/io/gravitee/gateway/reactive/flow/AbstractFlowResolver.java [hunk 4]
```diff
@@ -35,4 +45,49 @@ protected AbstractFlowResolver(ConditionFilter<Flow> filter) {
...
+                if (iterator.hasNext()) {
+                    currentPosition = iterator.next();
+                }
+            }
+            off = next + 1;
+            count++;
+        }
+
+        context.request().pathParameters().add(parameters.get(currentPosition), pathInfo.substring(off));
+    }
```
Comment (jhaeyaert) [inline gravitee-apim-gateway/gravitee-apim-gateway-flow/src/main/java/io/gravitee/gateway/reactive/flow/AbstractFlowResolver.java:92]: This is probably something we could try to mutualize for both v2 and v4 to avoid duplicating this algorithm and ease the maintenance. wdyt?

File: gravitee-apim-gateway/gravitee-apim-gateway-flow/src/main/java/io/gravitee/gateway/reactive/v4/flow/AbstractFlowResolver.java [hunk 5]
```diff
@@ -35,4 +44,51 @@ protected AbstractFlowResolver(ConditionFilter<Flow> filter) {
...
+                context.request().pathParameters().add(parameters.get(currentPosition), pathInfo.substring(off, next));
+                if (iterator.hasNext()) {
+                    currentPosition = iterator.next();
+                }
+            }
+            off = next + 1;
+            count++;
+        }
+
+        context.request().pathParameters().add(parameters.get(currentPosition), pathInfo.substring(off));
```
Comment (jhaeyaert) [inline gravitee-apim-gateway/gravitee-apim-gateway-flow/src/main/java/io/gravitee/gateway/reactive/v4/flow/AbstractFlowResolver.java:92]: Same here. Having to maintain this algorithm at 3 different locations seems not ideal.

File: gravitee-apim-gateway/gravitee-apim-gateway-flow/src/test/java/io/gravitee/gateway/flow/AbstractFlowParametersProviderTest.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.mockito.Spy;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class AbstractFlowParametersProviderTest {
```
Comment (jhaeyaert) [inline gravitee-apim-gateway/gravitee-apim-gateway-flow/src/test/java/io/gravitee/gateway/flow/AbstractFlowParametersProviderTest.java:38]: I'm wondering if we could add some corner cases such as:

- `/`
- `/:id`
- `/:id/foo/bar`
- `/:id?foo=bar` (not sure for this one if it is allowed to add query param)
- `/::id` (not sure if having '::' is rejected or not and what would be the behavior on the gateway side)
- `/:id/foo/bar/:id` (not sure if we allow multivalued path param)
- `/:id` but multiple times to simulate multiple flows matching the request (not sure if path param id will a a list of N identical value or if it overrides the previous one)
- ....

File: gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/reactive/handlers/api/flow/resolver/ApiFlowResolver.java [hunk 7]
```diff
@@ -41,14 +43,15 @@ public ApiFlowResolver(Api api, ConditionFilter<Flow> filter) {
 
     @Override
     public Flowable<Flow> provideFlows(GenericExecutionContext ctx) {
-        return this.flows;
+        addContextRequestPathParameters(ctx, flows);
```
Comment (jhaeyaert) [inline gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/reactive/handlers/api/flow/resolver/ApiFlowResolver.java:46]: I'm not a fan of this approach as path parameters are added during the reactive chain assembly.
Also, I have some questions regarding the usage of path parameters and I'm wondering if we could find a smart approach to avoid populating them for nothing. Here are a few examples of things I have in mind:

- do not try to resolve path parameters if we know there is no flow with such `:param` in the the path definition
- do not try to resolve path parameters if we know there is no usage (do EL deferred variables can be applied?).

**Review by jhaeyaert — CHANGES_REQUESTED** [review]

File: gravitee-apim-definition/gravitee-apim-definition-model/src/main/java/io/gravitee/definition/model/v4/listener/http/HttpListener.java (hunk 3 above)
Comment (ytvnr) [reply gravitee-apim-definition/gravitee-apim-definition-model/src/main/java/io/gravitee/definition/model/v4/listener/http/HttpListener.java:70]: It's not directly related to the issue but it was to align with what was done in v2 definition with ApiDeserializer https://github.com/gravitee-io/gravitee-api-management/blob/b08ca348b3a10cd518e3238b8356d9769eb6cd8d/gravitee-apim-definition/gravitee-apim-definition-jackson/src/main/java/io/gravitee/definition/jackson/datatype/api/deser/ApiDeserializer.java#L215

Comment (ytvnr) [reply gravitee-apim-definition/gravitee-apim-definition-model/src/main/java/io/gravitee/definition/model/v4/listener/http/HttpListener.java:70]: to put in another PR, as it's another topic, it's confusing

**Comment by ytvnr** [comment]
needs more thinking on that topic
